### PR TITLE
KAFKA-15326: [5/N] Processing thread punctuation

### PR DESCRIPTION
**Pull request by lucasbru**
Implements punctuation inside processing threads. The scheduler
algorithm checks if a task that is not assigned currently can
be punctuated, and returns it when a worker thread asks for the
next task to be processed. Then, the processing thread runs all
punctuations in the punctionation queue.

Piggy-backed: take TaskExecutionMetadata into account when
processing records.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java [hunk 1]
```diff
@@ -87,11 +88,29 @@ private void runOnce(final long nowMs) {
             if (currentTask == null) {
                 currentTask = taskManager.assignNextTask(DefaultTaskExecutor.this);
             } else {
+                boolean progressed = false;
+
+                // First, attempt to punctuate
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: nit: I do not think we need this comment. It is clear from the code that we first punctuate.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java [hunk 2]
```diff
@@ -87,11 +88,29 @@ private void runOnce(final long nowMs) {
...
+                if (taskExecutionMetadata.canPunctuateTask(currentTask)) {
+                    if (currentTask.maybePunctuateStreamTime()) {
+                        log.trace("punctuated stream time for task {} ", currentTask.id());
+                        progressed = true;
+                    }
+                    if (currentTask.maybePunctuateSystemTime()) {
+                        log.trace("punctuated system time for task {} ", currentTask.id());
+                        progressed = true;
+                    }
+                }
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:110]: Is there a specific reason we punctuate before we process. In the current code path we first process and then punctuate.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java [hunk 3]
```diff
@@ -87,11 +88,29 @@ private void runOnce(final long nowMs) {
...
+
                 // if a task is no longer processable, ask task-manager to give it another
                 // task in the next iteration
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: This comment does not match the code it should describe. Can we please remove it?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java [hunk 4]
```diff
@@ -87,11 +88,29 @@ private void runOnce(final long nowMs) {
             if (currentTask == null) {
                 currentTask = taskManager.assignNextTask(DefaultTaskExecutor.this);
             } else {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: Why can we not process a newly assigned task immediately but we wait until the next iteration?

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @lucasbru !

Here my feedback.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java (hunk 4 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: We should be able to loop around very quickly here, and we'll have to duplicate the `currentTask != null` check... but okay if you prefer.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java (hunk 3 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java (hunk 2 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:110]: Not really. It should not matter but it's good to be consistent here.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java (hunk 1 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java]: done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java [hunk 5]
```diff
@@ -86,12 +87,29 @@ private void runOnce(final long nowMs) {
 
             if (currentTask == null) {
                 currentTask = taskManager.assignNextTask(DefaultTaskExecutor.this);
-            } else {
-                // if a task is no longer processable, ask task-manager to give it another
-                // task in the next iteration
-                if (currentTask.isProcessable(nowMs)) {
+            }
+
+            if (currentTask != null) {
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:92]: Just throwing ideas here.
Can we not move those tests into the `assignNextTask()` method, so that we guarantee that `assignNextTask()` never returns null. But probably, we also need to consider the case where no new task can be assigned.
Shouldn't we use a condition variable or similar in the case where a task cannot be assigned to avoid busy waiting?

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:92]: Hmm, that's a good point! Not sure if I'm worried about getting back null (we still need the check even with a condition variable), but it'd be good to avoid busy waiting. But we still want to shut-down cleanly, so we still want to check isRunning and isPaused as well...)

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:92]: Let's discuss that and put it into a separate PR. WDYT?

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:92]: Yes, I think this is good for a separate PR. Also not really related to punctuation anymore.

Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutor.java:92]: I checked this out and we need to be a bit careful here to not have the task executors to sleep when work is available. We had several bugs like this in state updater. We have to make sure to signal on the condtion variable any time a task becomes "processable". Here are some situations where a task becomes processable:

 - Task is unassigned from another TaskExecutor.
 - Task state is changed (should only happen inside when a task is locked inside the polling phase).
 - When tasks are unlocked.
 - When tasks are added.
 - New records available.
 - A task is resumed.

So in summary, we 

 - we should probably lock tasks when they are paused and unlock them when they are resumed. This belongs to the StreamThread integration work.
 - need to awake the task executors `DefaultTaskManager.unassignTask`, `DefaultTaskManager.unlockTasks`, `DefaultTaskManager.addTask`. 
 - awake the task executors after every polling phase.

**Review by cadonna — APPROVED** [review]
Thanks for the updates!

LGTM!

**Comment by cadonna** [comment]
Build failures are unrelated:
```
Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()
Build / JDK 11 and Scala 2.13 / kafka.network.DynamicConnectionQuotaTest.testDynamicIpConnectionRateQuota()
Build / JDK 11 and Scala 2.13 / kafka.log.remote.RemoteIndexCacheTest.testClose()
Build / JDK 11 and Scala 2.13 / kafka.network.ConnectionQuotasTest.testNonDefaultConnectionCountLimitAndRateLimit()
Build / JDK 11 and Scala 2.13 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()
Build / JDK 20 and Scala 2.13 / org.apache.kafka.clients.consumer.CooperativeStickyAssignorTest.testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout(boolean).hasConsumerRack = false
Build / JDK 20 and Scala 2.13 / org.apache.kafka.clients.consumer.StickyAssignorTest.testAssignmentAndGroupWithNonEqualSubscriptionNotTimeout(boolean).hasConsumerRack = false
Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()
Build / JDK 20 and Scala 2.13 / kafka.api.TransactionsTest.testBumpTransactionalEpoch(String).quorum=kraft
Build / JDK 20 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
Build / JDK 20 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
Build / JDK 20 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
```
